### ISSUE TITLE
Discussion of Scoped Mixins API

### ISSUE THREAD

**Issue by likern** [vuejs/rfcs#60]
- Start Date: 2019-06-24
- Target Major Version: 3.x
- Reference Issues: (fill in existing related issues, if any)
- Implementation PR: (leave this empty)

# Note
I opened this Issue to discuss **ideas** of scoped mixins API and gather ideas and feedback about it. The main idea is, instead of completely change the way how components are written, update mixins semantic to make it more composable and reusable.

# Summary

Unify logic-composition patterns under scoped mixins.

# Basic example
``` vue
export default {
  data() {
    return {
      value: 0
    };
  },
  methods: {
    increment() {
      this.value++
    }
  },
  computed: {
    plusOne: function () {
      return this.value + 1
    }
  },
  watch: {
    value: function (newVal, old) {
      console.log(`count * 2 is ${newVal * 2}`)
    }
  },
  mounted() {
    console.log(`mounted`)
  }
}
```

``` vue
<template>
  <div>
    <span>count is {{ count.value }}</span>
    <span>plusOne is {{ count.plusOne }}</span>
    <button @click="increment">count.value++</button>
  </div>
</template>

<script>
import Count from "@/mixins/Count";

export default {
  mixins: {
    count: Count
  },
  data() {
    return {
    };
  }
}
</script>
```

# Motivation

## Logic Composition

One of the key aspects of the component API is how to encapsulate and reuse logic across multiple components. With Vue 2.x's current API, there are a number of common patterns we've seen in the past, each with its own drawbacks. These include:

- Mixins as array of mixin objects (via the `mixins` option)
- Higher-order components (HOCs)
- Renderless components (via scoped slots)

### Mixins drawbacks

Mixins in its current state have several major drawbacks making it difficult to reason about them and compose.

 - **One of the main complains** about mixins is it's difficult to figure out sources the properties come from. When several mixins are used it becomes problematic to understand *which mixin contains property* and *how properties are grouped by mixins*.
- Namespace clashing. Mixins can potentially clash on property and method names.
- Unexpected merging strategy for mixins
- Order of mixins in array is important and mixins can overwrite each other's functionality

``` vue
<template>
  <h1>Problem 1 with mixins</h1>
</template>

<script>
// Mixins are composed in unexpected way
// Why is called only secondMixin and why it is called twice?
const firstMixin = {
  created: function() {
    this.hello();
  },
  methods: {
    hello: function() {
      console.log("I'm from first mixin!");
    }
  }
};

const secondMixin = {
  created: function() {
    this.hello();
  },
  methods: {
    hello: function() {
      console.log("I'm from second mixin!");
    }
  }
};

export default {
  name: "problem1",
  mixins: [firstMixin, secondMixin]
};
</script>
```

The new scoped-mixins API presents a clean and flexible way to compose logic inside and between components without any of these drawbacks. This can be achieved by extracting code related to a piece of logic into what we call a "scoped mixin" which can be easily injected into component. Here is an example of using a *scoped mixin* to extract the logic of listening to the mouse position:

``` js
export default {
  data() {
    return {
      x: 0,
      y: 0
    };
  },
  methods: {
    update(e) {
      this.x = e.pageX
      this.y = e.pageY
    }
  },
  mounted() {
    window.addEventListener('mousemove', this.update)
  },
  unmounted() {
    window.removeEventListener('mousemove', this.update)
  }
}
```

``` vue
// in consuming component

<template>
  <div>{{ mouse.x }} {{ mouse.y }} {{ other.z }}</div>
</template>

<script>
import mouseMixin from "./mixins/mouse"
import otherMixin from "./mixins/other"

export default {
  name: "app",
  mixins: {
    mouse: mouseMixin,
    other: otherMixin
  }
};
</script>
```

Note in the example above:

- Properties exposed to the template have clear sources since they are nested object properties, where *parent object is used as namespace*;
- Mixin objects can be registered at arbitrary name (used as namespace) so there is no namespace collision;
- There are no unnecessary component instances created just for logic reuse purposes.

## Type Inference

TypeScript type inference support *is out of scope* of this RFC. This RFC solely focuses on better logic composability / reusability and tries to be better mixins / HOC / renderless components alternative.

**Comment by posva** [comment]
**TLDR**: With the function api, any possible naming conflict can be fully controlled and overridden. The typescript inference would be a problem though, making this solution far inferior to the function api in terms of Development experiences, so to me it doesn't make sense to add this new API (even though the surface is very small) because it suffers from the same problems.

The main complain you mention (where do things come from) remains unsolved with the namespacing. At the end, namespacing will only solve name clashing, all other problems still remain because they are inherent to mixins themselves and how they are designed to work

To me, this namespacing can be achieved by having a function (very common behavior for customizing mixins):

```js
export function mouseTracking(namespace = 'mouse') {
  const x = namespace + 'X'
  const y = namespace + 'Y'
  const listener = Symbol()

  return {
    data: => ({ [x]: 0, [y]: 0 }),
    mounted() {
      this[listener] = ({ pageX, pageY }) => {
        this[x] = pageX
        this[y] = pageY
      }
      window.addEventListener('mousemove', this[listener])
    },
    unmounted() {
      window.removeEventListener('mousemove', this[listener])
    }
}
```

it does fail as well in type inference though, which to be fair cannot be ignored.

about 

> // Why is called only secondMixin and why it is called twice?

It's because methods override each other but created hooks stack and get called in order (you can search for merging strategy in docs). So only one of the methods is taken into account

**Comment by likern** [comment]
> The main complain you mention (where do things come from) remains unsolved with the namespacing. At the end, namespacing will only solve name clashing, all other problems still remain because they are inherent to mixins themselves and how they are designed to work
The main complain you mention (where do things come from) remains unsolved with the namespacing. At the end, namespacing will only solve name clashing, all other problems still remain because they are inherent to mixins themselves and how they are designed to work
I don't understand this. Could you, please, clarify? What **exacty** is not solved with namespacing? And what can't be controlled?

> any possible naming conflict can be fully controlled and overridden
The same with scoped mixins, because you control how scoped mixins are registered and at what namespace. They are not expected to be overwritten by each other, they are not merged into Vue component directly, scoped mixins can't have arbitrary access to Vue component (only with special restrictive context)

**Comment by privatenumber** [comment]
As @posva said, it looks like the function API solves all the problems you highlight and even more. I'm interested to hear your perspective on the function API in comparison to your scoped mixin proposal --how is this better?

**Comment by posva** [comment]
The namespacing (which is what you call scoped mixins, but it really is namespacing by wrapping in an object) doesn't solve 2 of 3 of the points (the unexpected merging strategy was a question, but it really is expected and can even be customized!):

- One of the main complains about mixins is it's difficult to figure out sources the properties come from. When several mixins are used it becomes problematic to understand which mixin contains property and how properties are grouped by mixins -> **this remains the same as it's completely implicit what properties exist inside of the `count` object (taking it from one of your examples)**
- Order of mixins in array is important and mixins can overwrite each other's functionality -> **they can still clash for props (because it cannot be `<comp :mixin.prop="..."/>` that would be a modifier) so props will still override, and still merge for hooks like `created`**

Mixins are not going away and aren't replaced by the function api anyway (thinking about props, which cannot be added with the function api)

**Comment by likern** [comment]
> this remains the same as it's completely implicit what properties exist inside of the count object (taking it from one of your examples)

Ok, let's look at @yyx990803 proposal:

``` js
function useMouse() {
  const x = value(0)
  const y = value(0)
  const update = e => {
    x.value = e.pageX
    y.value = e.pageY
  }
  onMounted(() => {
    window.addEventListener('mousemove', update)
  })
  onUnmounted(() => {
    window.removeEventListener('mousemove', update)
  })
  return { x, y }
}

// in consuming component
const Component = {
  setup() {
    const { x, y } = useMouse()
    const { z } = useOtherLogic()
    return { x, y, z }
  },
  template: `<div>{{ x }} {{ y }} {{ z }}</div>`
}
```
As **Functional API** claims to be composable and reusable, `useMouse()` function will be definitely imported into this component from other file.
I have a few **questions**: 
1. How do you know what is returned from `useMouse()` without looking at external file?
2. How do you know what variables to destructure from object, returned from `useMouse()`?
3. How it differs from that you don't know beforehand what `Count` mixins contains?

> they can still clash for props (because it cannot be <comp :mixin.prop="..."/> that would be a modifier) so props will still override, and still merge for hooks like created

And they **shouldn't be overrideable outside of the component** - it is part of data encapsulation principle. The same way you **can't override props passed to component children** from outside of component like  `<comp :child.prop="..."/>`. This breaks whole encapsulation.

When it comes to 

> so props will still override
how is it possible, if component fully controls properties - which it accepts, at what name and type? 

Component can easily register mixin as another name, if that clash happens - and this will be done automatically with IDE support.

Scoped mixins **do not have access** to internals of Vue component. It is not possible to modify props in Vue component, it's own `data`, etc. And they do not merge so that you no more can understand what is called.

Let's look at official documentation example:
``` js
var mixin = {
  methods: {
    foo: function () {
      console.log('foo')
    },
    conflicting: function () {
      console.log('from mixin')
    }
  }
}

var vm = new Vue({
  mixins: [mixin],
  methods: {
    bar: function () {
      console.log('bar')
    },
    conflicting: function () {
      console.log('from self')
    }
  }
})

vm.foo() // => "foo"
vm.bar() // => "bar"
vm.conflicting() // => "from self"
```
That's a problem - **after mixin merge, you don't know which function is called. From mixin, or from component?** To understand that you have both to look at source code of component **and** mixin.

_Here is scoped mixins:_
``` js
var mixin = {
  methods: {
    foo: function () {
      console.log('foo')
    },
    conflicting: function () {
      console.log('from mixin')
    }
  }
}

var vm = new Vue({
  mixins: {
     mixin: mixin
  },
  methods: {
    bar: function () {
      console.log('bar')
    },
    conflicting: function () {
      console.log('from self')
    }
  }
})

vm.foo() // => "foo" // Error! Mixins are not merged directly to root component
vm.bar() // => "bar" // Easy to see where it comes from
vm.conflicting() // => "from self"
vm.mixin.conflicting() // Easy to see where it comes from
```

**Comment by backbone87** [comment]
> How do you know what is returned from useMouse() without looking at external file?

Documentation and typing. Documentation via JSDoc doesnt work too well with the options based API, but that is reasonably solveable by extra tooling. The dealmaker is typing and developing a typing system "just for Vue option API" is beyond reasonable cost. There are 2 JS typing systems (TS and flow), both backed and driven by 2 of the largest software companys with considerable investments. So reusing/supporting one or both of them is the only road available. And imho TS "won" (correct me if i am wrong, but didnt facebook even discontinued flow?), so this decision was already made for us.

> How do you know what variables to destructure from object, returned from useMouse()?

Its basically the same question as above: Documentation and typing.

**Comment by LinusBorg** [comment]
Your namespaced `count` object is definitely an improvement over normal mixins insofar as it shows which pieces of data came from the same mixin.

> How it differs from that you don't know beforehand what Count mixins contains?

It's an inherent principle of modularized code that we need to be aware of the API they expose once we import them somewhere and make use of them. So that's not the point.

The point is that *once we have looked that up*, we can destructure the exported props from a function so we clearly can see everything that is coming from this functon:

```js
const { x, y } = useMouse()
```

And everytime we want to know all of the stuff that came from this mixin, we can come back to this one line and find it all right there.

For a "scoped mixin", we don't get this information in such a compact form. We would have to scan all of the component's code for occurrences of `this.mouse.*` to find out all of the properties that the `mouse` object from the mixin provides (and that we actually make use of).

So it's better than normal mixins in that we can see which properties came from where since the scoped mixin groups them in one object, agreed. But the information what properties this object actually contains is still scattered throughout the component, which is a small disadvantage over the function-based API in my book.

**Comment by posva** [comment]
About the 3 questions: explicit and automatic typings. Typing inference with mixins can be achieved for static mixins for functions returning mixins, the user would have to. With functions, it's almost automatic inference thanks to typescript even if you are doing javascript: your editor will figure it out, improving discoverability 


> And they shouldn't be overrideable outside of the component - it is part of data encapsulation principle. The same way you can't override props passed to component children from outside of component like <comp :child.prop="..."/>. This breaks whole encapsulation.

I think I wasn't clear, I'm saying that props will still get overridden when clashing into multiple namespaces/_scoped_ mixins unless the outside name is prefixed and camel cased like `count.value` => `countValue`

It's absolutely possible to define props through a mixin, you can define any option, even custom ones, so it's something that should be taken into consideration when thinking about a namespaced version of mixins

> To understand that you have both to look at source code of component and mixin

Exactly, but the same happens with namespaces mixins because it still is implicit like regular mixins

**Comment by Meekohi** [comment]
I asked a similar question in Discord:

[7:25 PM] meekohi: Q: Has anyone ever considered explicitly namespacing mixins, to solve both the typescript and namespace problems? I'm not deep enough to know what might go wrong but curious about:
```
  mixins: {
    animal: FlyingMammals,
    sport: WoodenSticks
  },
  computed: {
    pet_name(): {
      return this.animal.bat.name // namespaced data/prop
    },
    weapon(): {
      return this.sport.bat // namespaced data/prop
    }
  }
```
[7:26 PM] meekohi: I would think it's much easier to pass type information along if you can reuse the mixin's type definitions verbatim (vs trying to merge it into the component's, which seemed like the problem Class API ran into?)
[11:15 AM] gusto: @meekohi so here's what I got as a response:

In that case, I think it would be better to just instantiate mixin instance in data hook.
```
Vue.extend({
  data() {
    return {
      animal: new FlyingMammals(),
      sport: new WoodenSticks()
    }
  },
  // ...
})
```

Type inference of namespaced mixin may work well but that means we will need another type parameter to Vue interface and component options. We already have many type parameters and it complicates our type declaration. https://github.com/vuejs/vue/blob/dev/types/options.d.ts#L68-L73

So I guess it would be difficult to extend our typing with that kind of approach unfortunately...
In fact, there are PRs which is pending now because of this problem.
- https://github.com/vuejs/vue/pull/9732
- https://github.com/vuejs/vue/pull/9369

**Comment by boonyasukd** [comment]
> * How do you know what is returned from useMouse() without looking at external file?
> * How do you know what variables to destructure from object, returned from useMouse()?
> * How it differs from that you don't know beforehand what Count mixins contains?

The answer is full 100% type inference support: in a function-based API that vue team proposed, in a capable editor like VSCode, you just press `Ctrl + Space` inside `{ }` to have an auto-complete window listing all of the exported elements from `useMouse()` (in this case, `x` and `y`). If `useMouse()` happens to export a function, such function will also be shown in an auto-complete window, and selecting it with an arrow key should show its complete method signature. And with an upcoming version Vetur extension, AFAIU type inference will work all the way to `<template>` section of SFC as well. So, with a function-based API, you can definitely know what's returned **without** looking at an external file. This is the main reason why people are enthusiastic about full typing support: it allows you to discover API on-the-fly in the way statically typed language developers have enjoyed for years.

Also, in the case of namespace clashing, the reason why people like function-based API is because you can [always reassign variables](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names) during destructure should variable names ever collide. Thus you have a flexibility in your hand to rename variables to maximize code readability:

``` js
// in a consuming component

const Component = {
  setup() {
    // suppose each useXXX() returns a 'state' variable;
    // we simply rename them in the way that ease maintenance/reduce confusion
    const { state: mouseState } = useMouse()
    const { state: otherState } = useOther()
    return { mouseState, otherState }
  },
  // here we use 'mouseState.XXX', 'otherState.XXX', similar to your 'mixin.XXX'
  template: `<div>{{ mouseState.x }} {{ mouseState.y }} {{ otherState.z }}</div>`
}
```

I know that the original RFC doesn't do a great job explaining this, but this is definitely what it meant when it says: 
> Returned values from composition functions can be arbitrarily named so there is no namespace collision;

**Comment by yyx990803** [comment]
Aside from type inference support, another important capability this proposal lacks is the ability for mixins to interact with one another.

With the scoped approach you are suggesting, scoped mixins are **completely isolated from each other** - there is no way for one mixin to somehow leverage the state of another mixin, because they won't know under what name the other mixin is going to be registered by the consuming component.

With function-based API this is as simple as passing the returned state of one composition function into another:

``` js
export default {
  setup(props) {
    const valueA = useFeatureA(props)
    const valueB = useFeatureB(valueA)
  }
}
```

Note there doesn't need any implicit coupling between the two features because composition happens in the consuming component.

**Comment by marceloadsj** [comment]
To solve thia, mixins should be a function, like data, and the mixins instantiated:

```javascript
mixins() {
  const count = new Count();
  const mouse = new Mouse(count);

  return { count, mouse };
}
```

But what I see is, using the functional RFC we can easily implement a "useMixin":

```javascript
setup() {
  return {
    count: useMixin(Count)
  };
}
```

But with the mixin scoped RFC, you can't implement the fn API. This means, the functional API is more flexible, and provides mechanisms to support a bunch of use cases, even the ones we are not thinking now.

Well, this is what I think. I'm a heavily React user but the way the fn API was proposed here, for me, solve even some React Hooks problems I'm facing today, maybe I can create something like useNewVueApi() for React, hahaha!

**Comment by martinsotirov** [comment]
As much as I dislike mixins, this RFC seems a lot saner than completely changing the API by reimplementing React hooks, so full support from me for scoped mixins, even if I wouldn't use them.

**Comment by skyrpex** [comment]
> As much as I dislike mixins, this RFC seems a lot saner than completely changing the API by reimplementing React hooks, so full support from me for scoped mixins, even though I would never use them.

The two APIs can live together so there's no reason to use a solution that solves less problems, IMHO.

**Comment by enekesabel** [comment]
When I first fell into the mixin hell, I tried to solve the problematic use of mixins by refactoring my code to  use composition instead of inheritance (using multiple mixins felt similar to multiple inheritance).
I instantiated mixins programmatically and stored them as data in the parent component. It was quite a hacky solution, (passing props to the mixin via propsData, and setting 'this' as the parent component) , but it was a great design improvement. The main pain points of this approach were that propsData is not reactive, and lifecycle methods had to be handled manually.
I think the little example @marceloadsj provided could a good start, or we could be using something like this:
`
const FooMixin = Vue.extend({
  props: ['foo'],
})

export default {
  data: ()=>({
    foo: 1,
    bar: 2
  }),
  mixins: {
    fooMixin: FooMixin({foo: this.foo}) // or new FooMixin({foo: this.foo})
  }
}
`

I see that the functional api is more flexible, but I think something like this would be a good alternative if you don't want to mix 'setup' into your component.

**Comment by marceloadsj** [comment]
@enekesabel do you think there's place for both approach?
Maybe separated pkgs, or one main approach and another with a plugin.

```javascript
Vue.use(VueMixins);
// or maybe
Vue.use(VueFunctional);
```
A kind of self-service way to prepare your approach to work with the fw.
The downside is the complexity that can bring to newcomers!

**Comment by enekesabel** [comment]
@marceloadsj I don't know, I think that would be up to the core members to decide.

Imo calling this feature mixins/scoped mixins is misleading since nothing would be "mixed" into the component definition. This is just a more conscious approach which would let us reuse components without the shortcomings of mixins.

**Comment by likern** [comment]
> @marceloadsj I don't know, I think that would be up to the core members to decide.
> 
> Imo calling this feature mixins/scoped mixins is misleading since nothing would be "mixed" into the component definition. This is just a more conscious approach which would let us reuse components without the shortcomings of mixins.

I agree, it might be misleading. Do you have any ideas of how to name it?
I wanted "scoped mixins" to be based to mixins ideas. But the main point - solve problems which tries to solve **Functional API**, but in context of Object API. So it will be improved way to compose and add functionality into component.

It will give ability to gradually improve and refactor component, which becomes more verbose and complex. This is the way how developers usually handle with complexity. They start with simple components... complexity and functionality is increasing... and after that some part refactor out into another component or mixin to become shareable. It this point **I don't expect develops will say**: "Ouh...that component is too big and complex! Let's remove all code and rewrite with _Functional API_".

_We need something improved, better alternative than current mixin state, something what doesn't unexpectedly mixin / overwrite functionality of component._

**Comment by yyx990803** [comment]
@likern the major problem I see with this proposal is that, as pointed out by previous comments, it solves strictly fewer problems than the function-based API: no type inference, no ability to interact between mixins. It is an improvement over current mixins for sure, but it is also a step back from the function API.

> It this point I don't expect develops will say: "Ouh...that component is too big and complex! Let's remove all code and rewrite with Functional API".

You **can** gradually migrate your options-based logic into the function API. You don't have to rewrite the whole thing:

``` diff
export default {
  // replace data() with setup()
  setup() {
+   const { c1, m1, baz } = useFeature()
    return {
      foo: 1,
      bar: 2,
-     baz: 3
+     baz,
+     c1,
+     m1
    }
  },
  computed: {
-   c1() { /* ... */ },
    // other computed
  },
  methods: {
-   m1() { /* ... */ },
    // other methods
  }
}
```

**Comment by likern** [comment]
> Your namespaced `count` object is definitely an improvement over normal mixins insofar as it shows which pieces of data came from the same mixin.
> 
> > How it differs from that you don't know beforehand what Count mixins contains?
> 
> It's an inherent principle of modularized code that we need to be aware of the API they expose once we import them somewhere and make use of them.
> 
> The difference is that _once we have looked that up_, we can destructure the exported props from a function so we clearly can see everything that is coming from this functon:
> 
> ```js
> const { x, y } = useMouse()
> ```
> 
> And everytime we want to know all of the stuff that came from this mixin, we can come back to this one line and find it all right there.
> 
> For a "scoped mixin", we don't get this information in such a compact form. We would have to scan all of the component's code for occurrences of `this.mouse.*` to find out all of the properties that the `mouse` object from the mixin provides (and that we actually make use of).
> 
> So it's better than normal mixins in that we can see which properties came from where since the scoped mixin groups them in one object, agreed. But the information what properties this object actually contains is still scattered throughout the component, which is a small disadvantage over the function-based API in my book.

@LinusBorg **Functional API** was presented as a way to solve complex problems, for experienced developers and advanced usage. ES6 Object destructuring wasn't invented for that case, when you have multiple components.

As soon as complexity of component is increasing - you will get a lot of `const {x, y} = useOneThing()` statements. With high probability of name clashes. **These object props are never named to have enough information to be distinguishable on its own**, because they are expected to be used with object prefix.
``` js
const { signIn } = useFirebaseAuth()
```
it's almost never named like this
``` js
const { firebaseSignIn } = useFirebaseAuth()
```

So I have to state that it's not a valid point. **The more complex component - the more you need namespace**. With *scoped mixins* **you already have that namespace which you already have chosen**. This
``` js
const { signIn } = useGithubAuth()
const { signIn } = useTwitterAuth()
```
becomes this somewhere in code
``` js
githubAuth.signIn() // or this.githubAuth.signIn()
twitterAuth.signIn() // this.twitterAuth.signIn()
```

So, for complex components, for which Functional API was proposed you will get
``` js
const githubAuth = useGithubAuth()
const twitterAuth = useTwitterAuth()
...
githubAuth.signIn()
twitterAuth.signIn()
```
it's no difference with *scoped mixins syntax*

> For a "scoped mixin", we don't get this information in such a compact form.
Agree. But, on the other hand - this is only a small, negligible disadvantage. You always can look into mixin object to see API or to get autocompletion (at least theoretically, not sure about that). Just looking on object destructing on its own doesn't give valuable information.

**But**, with object destructuring refactoring **becomes more complex**, and not automatic.
``` js
setup() {
    const { signIn } = useGitHubAuth() // also has signUp
    const { signUp } = useTwitterAuth() // also has signIn
    return {
        signIn, signUp
    }
}
```

How do you know inside template from which `useSomething()` function comes from?
How do you easily refactor that code? With *scoped mixins* you can do that with command "Change All Occurences" and *all* mixin methods will be under another prefix.

With Functional API this is not possible, because:

1. No easy way to understand where function comes from (from which `useFeature()`). Again, the same problem as old mixins
2. Properties can awkwardly be renamed
``` js
setup() {
    const { signIn } = useGitHubAuth()
    const { signUp: githubSignIn } = useTwitterAuth()
    return {
        signIn, githubSignIn
    }
}
```

**You have to go through all the methods you have destructured and rename them manually**. Lots of work!

**Comment by yyx990803** [comment]
> How do you easily refactor that code?

Changing symbol names of functions and variables are the basic features of IDEs today. In VSCode you can just right click and "rename symbol" - that's it.

**Comment by likern** [comment]
> Aside from type inference support, another important thing this proposal misses is the ability for mixins to interact with one another.
@yyx990803 This is a valid point, but it's only a first discussion. Definitely interaction between mixins can be solved.

**Comment by marceloadsj** [comment]
This is a simple example of what can be done using the functional API (made for React, but I think can be replicated to Vue as well):

```javascript
class Count {
  state = { value: 0 };

  increment = () => this.setState({ value: this.state.value + 1 });

  // React dont have computed, so using plaing fn
  plusOne = () => this.state.value + 1;

  // This is how we can run code when some state change, React dont have watch
  componentDidUpdate = (_, prevState) => {
    if (prevState.value !== this.state.value) {
      console.log(`count * 2 is ${this.state.value * 2}`);
    }
  };

  componentDidMount = () => console.log(`mounted`);
}

function MyCounter(props) {
  const count = useMixin(Count, props);

  return (
    <div>
      <p>count is {count.state.value}</p>
      <p>plusOne is {count.plusOne()}</p>
      <button onClick={count.increment}>count.value++</button>
    </div>
  );
}
```

And the useMixin hook is:
```javascript
function useMixin(Mixin, props) {
  const [prevProps, setPrevProps] = React.useState();
  const [prevState, setPrevState] = React.useState();

  const mixin = React.useMemo(() => {
    const instance = new Mixin();

    instance.setState = newState => {
      setPrevState(instance.state);
      instance.state = { ...instance.state, ...newState };
    };

    return instance;
  }, []);

  React.useEffect(() => {
    if (mixin.componentDidMount) mixin.componentDidMount();
    if (mixin.componentWillUnmount) return mixin.componentWillUnmount();
  }, []);

  React.useEffect(() => {
    if (mixin.componentDidUpdate && (prevProps || prevState)) {
      mixin.componentDidUpdate(prevProps, prevState);
    }
  }, [prevProps, prevState]);

  return mixin;
}
```

In my opinion, the flexibility is amazing, and we can even continue to work the way we do today (just with really small changes on some situations).

Obs: sorry to use React always but, it's just to illustrate the point and because the Hooks and the fn API is likely the same (for me the Vue approach is even easier to read)...

**Comment by nekosaur** [comment]
> As soon as complexity of component is increasing - you will get a lot of `const {x, y} = useOneThing()` statements. With high probability of name clashes. **These object props are never named to have enough information to be distinguishable on its own**, because they are expected to be used with object prefix.
> 
> ```js
> const { signIn } = useFirebaseAuth()
> ```
> 
> it's almost never named like this
> 
> ```js
> const { firebaseSignIn } = useFirebaseAuth()
> ```
> 

But no one is forcing you to destructure. You can just as well do

```js
const github = useGithubAuth()
const twitter = userTwitterAuth()

return {
  github,
  twitter
}
```

which gives you your namespaces.

**Comment by likern** [comment]
@marceloadsj That's what I talked about. It is very difficult to understand what is happening in your code. Difficult to reason about and maintain. Looks like hodgepodge. After several months you will try hard to understand how that works and what is happening.

Objects API is **declarative**. Template syntax is **declarative**.
**React is not** - it is difficult to understand what is exactly will be rendered, because it generates JSX dynamically in functions, based on arbitrary conditions (this is how it's used mostly)


That is why Vue templates are far superior React approach. It is why developers have chosen Vue over React. It is not aesthetic / beautiful look. It's all about **readability, maintainability**.

But React is more powerful - you can do any computations and generate arbitrarily complex JSX. But have you heard Vue users ever complained about it? No, never. **Because that flexibility is disadvantage**.

The same will happen with **Functional API**. Is it more flexible than Object API? Yes. Have you seen people ever complained about Object API and that React Hooks gives more? I've never heard that. Because it's considered **disadvantage**.

**Declarative approach is always more readable and simple to reason about and understand. Especially for complex components and projects.**

**Comment by likern** [comment]
> > As soon as complexity of component is increasing - you will get a lot of `const {x, y} = useOneThing()` statements. With high probability of name clashes. **These object props are never named to have enough information to be distinguishable on its own**, because they are expected to be used with object prefix.
> > ```js
> > const { signIn } = useFirebaseAuth()
> > ```
> > 
> > 
> > it's almost never named like this
> > ```js
> > const { firebaseSignIn } = useFirebaseAuth()
> > ```
> 
> But no one is forcing you to destructure. You can just as well do
> 
> ```js
> const github = useGithubAuth()
> const twitter = userTwitterAuth()
> 
> return {
>   github,
>   twitter
> }
> ```
> 
> which gives you your namespaces.

Yes, I can. Initially namespaces were proposed as disadvantage (like you can't destructure). But, eventually, we get the same in scoped mixins and Functional API for **complex components.**

No difference here between both.

**Comment by skyrpex** [comment]
> @marceloadsj That's what I talked about. It is very difficult to understand what is happening in your code. Difficult to reason about and maintain. Looks like hodgepodge. After several months you will try hard to understand how that works and what is happening.
I have to say that the Vue's functional API is much clearer compared to React Hooks... By far!

> 
> Objects API is **declarative**. Template syntax is **declarative**.
> **React is not** - it is difficult to understand what is exactly will be rendered, because it generates JSX dynamically in functions, based on arbitrary conditions (this is how it's used mostly)
> 
> That is why Vue templates are far superior React approach. It is why developers have chosen Vue over React. It is not aesthetic / beautiful look. It's all about **readability, maintainability**.
That's not my case. I choose Vue because of the state management (reactivity), but I prefer JSX and hooks to write my code.

> 
> But React is more powerful - you can do any computations and generate arbitrarily complex JSX. But have you heard Vue users ever complained about it? No, never. **Because that flexibility is disadvantage**.
IMHO, at the risk of sounding rude, it's a disadvantage only if you can't handle it. You can define your own rules over a flexible system if you really need it.

> 
> The same will happen with **Functional API**. Is it more flexible than Object API? Yes. Have you seen people ever complained about Object API and that React Hooks gives more? I've never heard that. Because it's considered **disadvantage**.
Me, every single day I use Vue.

> 
> **Declarative approach is always more readable and simple to reason about and understand. Especially for complex components and projects.**
I disagree...

**Comment by likern** [comment]
The same opinion was mentioned here https://github.com/vuejs/rfcs/pull/42#issuecomment-505447468. That is where Evan is wrong. It's not about aesthetic at all. It's about **declarative vs imperative** approach. Readability and maintainability.

**Comment by likern** [comment]
> @skyrpex IMHO, at the risk of sounding rude, it's a disadvantage only if you can't handle it. You can define your own rules or conventions over a flexible system if you really need it.

Because Javascript is not a language which can handle this flexibility. It is like plain C. It's not functional at all. It hasn't advanced type-system, doesn't have metaprogramming, it's not possible to create DSL.

**Comment by marceloadsj** [comment]
I'm not sure if I'm understanding correctly, it not you can just ignore my comments here, hahaha :smile:, but, well, for me it's not like you said:

I completely agreed with you about that the **hooks** is harder to read than the **obj/class** on React. useEffect, useCallback, useMemo, the names don't tell so much about what happen, forcing you to know all the documentation.

But looking at the Vue approach, it's way better because the functions are very descritive: onMount, onUnmount, watch, computed, everything make sense when you read line by line.

It seems that you are "sacrificing" a whole better way (better because it's composable, it's reusable, flexible, well, all the pros we already know) to solve some hard problems just because of that point.

One of the advantages of using that functional approach is that you can design your API (like the example I posted), and that's it, everything write on top of that code will be fully declarative, using the obj api, as we use today.

If the Vue team implement that approach in the core, we will never know that it uses the fn api. But for everyone that wants to work "closer to the metal", they can just use these fn api directly.

**Comment by skyrpex** [comment]
@likern I'm confused... First you mention that React is too flexible and it's a disadvantage, and now you say that Javascript is not flexible. Could you please rephrase?

**Comment by likern** [comment]
> > Because Javascript is not a language which can handle this flexibility. It is like plain C. It's not functional at all. It hasn't advanced type-system, doesn't have metaprogramming, it's not possible to create DSL.
> 
> @likern I'm confused... First you mention that React is too flexible and it's a disadvantage, and now you say that Javascript is not flexible. Could you please rephrase?

It's flexible because it is not Javascript. It's a whole new thing on top of it.

1. React is not truly a Javascript.
2. JSX is not a JavaScript and computations in it's properties too - it's syntax which mimics JS. And they adopt it to new ES6 features. Similar to Vue templates - not Javascript.

Javasciprt doesn't allow to create advanced abstractions. Doesn't allow to create JSX or similar syntax which will be valid Javascript and no more.

**Comment by likern** [comment]
Ok, I've understood the general attitude of fans of Functional API. I think, even if all mentioned problems will be resolved (like interaction between mixins) it will not be accepted by Vue team. So later I'll close this issue.

**Comment by yyx990803** [comment]
@likern I would say options-based API is only "declarative" at a superficial level. Inside a lifecycle hook, inside a method, inside a watcher callback, you are still writing imperative code. The only thing "declarative" about options-based API is that it declares the options. This superficial declarative-ness sacrifices flexibility - and not in a good way. It forces code belonging to the same logic topic to be split between different option types, and creates pretty much all the problems your proposal is trying so hard to deal with.

On the other hand, is imperative really bad? JavaScript is a pretty imperative language. If you want to go really "full declarative", you should be using a functional compile-to-js language like Elm or ReasonML. The fact that a lot of us are happy about JavaScript and are productive with it shows the declarative vs. imperative dichotomy is not so black vs. white.

**Comment by skyrpex** [comment]
> > > Because Javascript is not a language which can handle this flexibility. It is like plain C. It's not functional at all. It hasn't advanced type-system, doesn't have metaprogramming, it's not possible to create DSL.
> > 
> > 
> > @likern I'm confused... First you mention that React is too flexible and it's a disadvantage, and now you say that Javascript is not flexible. Could you please rephrase?
> 
> It's flexible because it is not Javascript. It's a whole new thing on top of it.
> 
> 1. React is not truly a Javascript.
> 2. JSX is not a JavaScript and computations in it's properties too - it's syntax which mimics JS. And they adopt it to new ES6 features. Similar to Vue templates - not Javascript.
> 
> Javasciprt doesn't allow to create advanced abstractions. Doesn't allow to create JSX or similar syntax which will be valid Javascript and no more.

I think you're confusing some terms. You can use React with pure Javascript, even in render functions. Now, rendering HTML using `h("div", { classNames: "text-red" })` can be cumbersome, and that's why JSX was created. Finally, JSX is just a superset of Javascript that get's transpiled to the code I mentioned above. Nothing more fancy than that (you can opt out).

> I think, even if all mentioned problems will be resolved (like interaction between mixins) it will not be accepted by Vue team.

I don't think that any sort of mixin based solution will solve all the current problems. It would look like the functional API at the end, I'm almost sure of it.
